### PR TITLE
os/bluestore/BlueFS: explicit check for too-granular allocations

### PR DESCRIPTION
**Pull request by liewegas**
Be helpful and suggest what config option to adjust.

It is not easy to restart the replay process with a finer granularity,
since replay is adjusting in-memory state as it goes.

Fixes: https://tracker.ceph.com/issues/43904 (sort of)
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
This is a minimal fix that requires human intervention, but I don't think these OSDs will be very common...

**Comment by liewegas** [comment]
/a/sage-2020-02-01_03:27:35-rados-wip-sage-testing-2020-01-31-1746-distro-basic-smithi/4723116 fails the qa/standalone/osd/osd-bluefs-volume-ops.sh

also fails vstart

**Comment by ifed01** [comment]
jenkins retest this please

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -336,24 +336,41 @@ int BlueFS::reclaim_blocks(unsigned id, uint64_t want,
...
-    log_t.op_alloc_rm(id, p.offset, p.length);
-  }
+  if (want > 0) {
+    got += alloc[id]->allocate(want, alloc_size[id], 0, extents);
+    ceph_assert(got != 0);
+    if (got < 0) {
+      derr << __func__ << " failed to allocate space to return to bluestore"
+	   << dendl;
+      alloc[id]->dump();
+      return got;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: we might leak extents from block_unused_too_granular at this point unless we push them back to the list

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -833,16 +834,27 @@ int BlueFS::_check_new_allocations(const bluefs_fnode_t& fnode,
...
+  __u8 id, uint64_t *offset, uint64_t *length, bool alloc)
 {
+  const char *op = alloc ? "op_alloc_add" : "op_alloc_rm";
   auto oldo = *offset;
   auto oldl = *length;
   if (*offset & (alloc_size[id] - 1)) {
     *offset &= ~(alloc_size[id] - 1);
     *offset += alloc_size[id];
     if (*length > *offset - oldo) {
+      if (alloc) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: You should process a tail left unused in the same way after doing
 *length &= ~(alloc_size[id] - 1);
below.

**Comment by ifed01** [comment]
Third commit to be squashed I guess.

File: src/os/bluestore/BlueFS.cc [hunk 3]
```diff
@@ -833,20 +834,36 @@ int BlueFS::_check_new_allocations(const bluefs_fnode_t& fnode,
...
+	block_unused_too_granular[id].insert(oldo, *length);
+      } else {
+	block_unused_too_granular[id].erase(oldo, *length);
+      }
       *length = 0;
     }
   }
-  *length &= ~(alloc_size[id] - 1);
+  if (*length & (alloc_size[id] - 1)) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: Shouldn't we process !alloc case and remove from the list here as well?

Comment (liewegas) [reply src/os/bluestore/BlueFS.cc]: yep!

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Pull request merged**